src/yarn.ts

```diff
@@ -8,99 +8,192 @@
+        // OK, now grab folders for each of these dependencies
         const nodePackages: INodePackage[] = [];
-        inputTrees.forEach((yarnNode: IYarnNode) => this.addNodePackage(nodeModulesFolder, yarnNode, nodePackages));
+        subsetDependencies.forEach(moduleName => this.addNodePackage(nodeModulesFolder, moduleName, nodePackages));
 
-        // return uniq entries
+        // return unique entries
         return Promise.resolve(nodePackages.map((e) => e.path).filter((value, index, array) => {
             return index === array.indexOf(value);
         }));
     }
 
     /**
-     * Add a node package (entry of yarn list) to the given array.
-     * Also loop on all children and call ourself back
-     * @param nodeModulesFolder the node_modules location
-     * @param yarnNode the node entry to add
-     * @param packages the array representing all node dependencies
+     * Find from children all the direct dependencies. Also exclude some dependencies by not analyzing them.
+     * Allow as well to report error in case of a forbidden dependency found
+     * @param children the list of dependencies to analyze
+     * @param nodeTreeDependencies the object containing the tree of dependencies
+     * @param subsetDependencies  the
      */
-    protected async addNodePackage(nodeModulesFolder: string, yarnNode: IYarnNode, packages: INodePackage[]): Promise<void> {
+    protected findDependencies(children: string[], nodeTreeDependencies: Map<string, string[]>, subsetDependencies: string[]): void {
+
+        children.map(child => {
+            // only loop on exist
+            if (subsetDependencies.indexOf(child) >= 0) {
+                return;
+            }
+            subsetDependencies.push(child);
+
+            // loop on children in any
+            let depChildren = nodeTreeDependencies.get(child);
+            if (depChildren) {
+                depChildren = depChildren.filter(depChild => {
+                    const res = this.excludedPackages.indexOf(depChild) < 0;
+                    if (!res) {
+                        Logger.debug(` --> Excluding the dependency ${depChild}`);
+                    }
+                    return res;
+                });
+
+                const matching: string[] = [];
+                const foundForbiddenPackage = depChildren.some(r => {
+                    const res = this.forbiddenPackages.indexOf(r) >= 0;
+                    if (res) {
+                        matching.push(r);
+                    }
+                    return res;
+                });
+                if (foundForbiddenPackage) {
+                    throw new CliError(`Forbidden dependencies ${matching} has been found as dependencies of ${child}` +
+                        `Current dependencies: ${depChildren}, excluded list: ${this.forbiddenPackages}`);
+                }
+                this.findDependencies(depChildren, nodeTreeDependencies, subsetDependencies);
+            }
+        }
+        );
```

Comment:
Maybe it should be upper in one line }); ? But up to you.

Reply:
fixed